rr_rviz_plugins/src/ChassisPanel.cpp

```diff
@@ -0,0 +1,198 @@
+    // does not affect what the label says
+    if (cur_speed >= top_speed) {
+        cur_speed = top_speed;
+    }
+    // the angle the speedometer needle needs to point in
+    auto angle = (((top_speed - cur_speed) * M_PI) / top_speed);
+    // determines x and y height of the needle
+    float x_comp = cos(angle) * width / 2;
+    float y_comp = sin(angle) * height / 2;
+    // the shape and direction of the needle
+    QLineF needle =
+          QLineF(width / 2 + x_pos, y_pos + height / 2 - 1, width / 2 + x_comp + x_pos, y_pos + height / 2 - y_comp);
+    painter.setBrush(Qt::red);
+    painter.setPen(Qt::red);
+    painter.drawLine(needle);
+    // new params for the heading arrow
+    y_pos = 38;
+    height = 18;
+    // determines the rectangle shape the arrow will fit in
+    rect = QRectF(x_pos, y_pos, width, height);
+    painter.setPen(QColor(255, 255, 255, 255));
+    painter.setBrush(QColor(255, 255, 255, 255));
+    painter.drawRect(rect);
+    painter.setBrush(Qt::black);
+    painter.setPen(Qt::black);
+    int x_org = x_pos + width / 2;
+    int y_org = y_pos + height / 2;
+    // The size of the arrow
+    int arr_size = 9;  // this is actually  half the length of the arrow
+    // makes it so that zero points north instead of east
+    hding = hding + M_PI / 2;
+    y_comp = sin(hding) * (float)arr_size;
+    x_comp = cos(hding) * (float)arr_size;
+
+    QLineF arrow_shaft = QLineF(x_org - x_comp, y_org + y_comp, x_org + x_comp, y_org - y_comp);
+    painter.drawLine(arrow_shaft);
+
+    // For the arrowhead
+    painter.setPen(Qt::red);
+    painter.setBrush(Qt::red);
+    QLineF arrow_tip = QLineF(x_org + x_comp, y_org - y_comp, x_org + .5 * x_comp, y_org - .5 * y_comp);
+    painter.drawLine(arrow_tip);
+
+    // For the status of recording_bag
+    x_pos = 137;
+    y_pos = 95;
+    width = 40;
+    height = 15;
+    if (received_signal && record_bag) {
+        painter.setBrush(Qt::green);
+        painter.setPen(Qt::green);
+        painter.drawRect(x_pos, y_pos, width, height);
+    } else if (received_signal) {
+        painter.setBrush(Qt::red);
+        painter.setPen(Qt::red);
+        painter.drawRect(x_pos, y_pos, width, height);
+    }
+
+    // for the status of estop
+    x_pos = 52;
+    y_pos = 122;
+    width = 23;
+    if (received_signal && estop) {
+        painter.setBrush(Qt::red);
+        painter.setPen(Qt::red);
+        painter.drawRect(x_pos, y_pos, width, height);
+    } else if (received_signal) {
+        painter.setBrush(Qt::green);
+        painter.setPen(Qt::green);
+        painter.drawRect(x_pos, y_pos, width, height);
+    }
+    QWidget::paintEvent(e);
+}
+
+void ChassisPanel::chassisStateCallback(const rr_msgs::chassis_state msg) {
+    // for timerCallback
+    chassisUpdateTime = ros::Time::now();
+    received_signal = true;
+
+    // global variables
+    cur_speed = msg.speed_mps;
+    hding = msg.steer_rad;
+    record_bag = msg.record_bag;
+    estop = msg.estop_on;
```

Comment:
Personally, I think the name like steer_rad is more clear than hding. Maybe just use the same names that chassis state uses